socs/agents/hi6200/agent.py

```diff
@@ -0,0 +1,137 @@
+import argparse
+import time
+
+from ocs import ocs_agent, site_config
+from ocs.ocs_twisted import TimeoutLock
+
+from socs.agents.hi6200.drivers import Hi6200Interface
+
+
+class Hi6200Agent:
+    def __init__(self, agent, ip_address, tcp_port):
+        self.agent = agent
+        self.log = agent.log
+        self.lock = TimeoutLock()
+
+        self.job = None
+        self.ip_address = ip_address
+        self.tcp_port = tcp_port
+        self.monitor = False
+
+        self.scale = None
+
+        # Registers Scale Output
+        agg_params = {
+            'frame_length': 10 * 60,
+        }
+        self.agent.register_feed('scale_output',
+                                 record=True,
+                                 agg_params=agg_params,
+                                 buffer_time=0)
+
+    @ocs_agent.param('_')
+    def init(self, session, params=None):
+        """init()
+
+        **Task** - Initialize connection to the Hi 6200 Weight Sensor.
+
+        """
+        with self.lock.acquire_timeout(0) as acquired:
+            if not acquired:
+                return False, "Could not acquire lock"
+
+            self.scale = Hi6200Interface(self.ip_address, self.tcp_port)
+
+            self.log.info("Connected to scale.")
+
+        return True, 'Initialized Scale.'
+
+    @ocs_agent.param('wait', type=float, default=1)
+    def monitor_weight(self, session, params=None):
+        """
+
+        **Process** - Continuously monitor scale gross and net weights.
+
+        Parameters:
+            wait (float, optional): Time to wait between measurements
+                [seconds].
+
+        """
+        session.set_status('running')
+        self.monitor = True
+
+        while self.monitor:
+            with self.lock.acquire_timeout(1) as acquired:
+                if acquired:
+                    data = {
+                        'timestamp': time.time(),
+                        'block_name': 'weight',
+                        'data': {}
+                    }
+
+                    try:
+                        data['data']["Gross"] = self.scale.read_scale_gross_weight()
+                        data['data']["Net"] = self.scale.read_scale_net_weight()
+
+                        self.agent.publish_to_feed('scale_output', data)
+
+                        # Allow this process to be queried to return current data
+                        session.data = data
+
+                    except ValueError as e:
+                        self.log.error(f"Scale responded with an anomolous number, ignorning: {e}")
+
+                    except AttributeError as e:
+                        self.log.error(f"Scale dropped TCP connection momentarily, trying again: {e}")
```

Comment:
You are correct. I believe I was confusing the AttributeError with a ValueError. Sometimes the scale will randomly respond with a None and break the hex decoding with a ValueError. I was attempting to catch these errors but got mixed up with scale becoming None... which would cause an AttributeError. I've swapped this to a ValueError.

Reply:
Maybe the `auto_open=True` is effectively handling this re-initialization?

It sounds to me like the `AttributeError` could still occur, so I'm not convinced this is resolved yet. I'd suggest, if `None` could come back from the register -- I'm assuming from `self.scale.read_holding_registers()` -- to handle that before trying to pass to `decode_scale_weight_registers`.